### ISSUE TITLE
[2025/05/17] 언어 모델은 어떤 구조로 사고하는 것일까?]

### ISSUE THREAD

**Issue by DaewooKim** [DaewooKim/AGI#3]
# [언어 모델은 어떤 구조로 사고하는 것일까?](https://www.linkedin.com/posts/gim-gyungjin-7430aab4_%EC%96%B8%EC%96%B4-%EB%AA%A8%EB%8D%B8%EC%9D%80-%EC%96%B4%EB%96%A4-%EA%B5%AC%EC%A1%B0%EB%A1%9C-%EC%82%AC%EA%B3%A0%ED%95%98%EB%8A%94-%EA%B2%83%EC%9D%BC%EA%B9%8C-anthropic%EC%97%90%EC%84%9C%EB%8A%94-%EC%B5%9C%EA%B7%BC-activity-7324709893137928192-iEyv?utm_source=share&utm_medium=member_desktop&rcm=ACoAABAfX0cBcPd9w9fX5dTNsijrnYhcRqaSjYc)

Anthropic에서는 최근 자사 블로그와 CEO 포스팅을 통해 AI의 작동 방식에 대한 이해가 중요하다는 메시지를 계속 전하고 있다.[1] 단순한 성능 향상이 아니라, 그 성능이 만들어지는 과정 자체를 이해하지 못하면 AI가 보여줄 수 있는 위험을 예측하거나 제어할 수 없기 때문이다.
이 분야의 경쟁이 치열하기 때문에 모델의 사고 수준은 예상보다 훨씬 빠르게 성장 중이다. (GPT-3.5의 추론 비용은 2년 사이에 280배나 줄었음[7]). 모델이 예상보다 빠르게 "사고" 능력을 진화시키는 만큼, 그 내부 사고 과정을 들여다보는 해석 연구가 반드시 필요하다는 문제 제기가 지속되고 있다. 일부 포스팅은 이 리스크를 체르노빌 사고에 비유하기도 하기도 하는데[2], agent 시스템이 발전하는 속도를 봤을때 근시일 내에 대형 보안 사고나 시스템 제어 사고등이 일어나지 말란 법은 없다.

초기 해석 연구에서는 뉴런을 직접 확인하여, 모델이 개념을 어떻게 이해하고 있는지 확인하려 했지만 이는 잘 동작하지 않았다. 그 이유는 뉴런 개수보다 표현해야 할 개념 수가 많기 때문. 하나의 뉴런이 여러 개념을 중첩하여 표현하고 있기 때문에 뉴런 단위로 해석하는 경우 인간이 알아보기 쉽지 않기 때문이다. 이를 해결하기 위해 Anthropic은 CLT(Cross-Layer Transcoder)라는 방법을 제시하고 있다.

- Circuit Tracing과 CLT
이 idea는 transformer도 결국 여러 MLP(Multi Layer Perceptron)의 조합인데, 이 레이어가 변화할 때마다, 어떤 변화가 발생하는지 감지할 수 있으면, 모델이 어떤생각을 하는지 알수 있다는 발상에서 나온 것이다.
CLT는 기존 transformer는 freeze 둔 상태로 CLT 모듈을 추가 학습 하는 형태다. 일종의 sparse encoder decoder를 붙여서 해석기를 만드는 과정이라 할 수 있는데, 크게 다음 2가지의 목적을 가진다.
1 loss term을 조정하여 의도적으로 소수의 feature만 활성화되도록 유도함. 이를 통해 다음 레이어에서 어떤 feature가 의미 있는 역할을 하는지 추적 가능.
2 일반적인 뉴런보다 더 많은 표현 공간을 사용하므로, 각 feature는 더 구체적이고 명확한 의미를 갖게 됨.

예를 들어 어떤 문장 안에서 모델이 "프랑스의 수도는 파리이다"라는 문장을 생성하려고 할 때를 생각해 보자.
초반에는 "프랑스"라는 단어를 보고 모델은 국가 개념에 대한 feature를 활성화할 것이다. 이어서 "수도는"이 등장하면, CLT는 "국가 → 수도"라는 관계 feature를 감지할 수 있다. 마지막으로 "파리"를 출력하기 직전, CLT에서 "프랑스의 수도 = 파리"와 관련된 특징들이 강하게 활성화되면서 출력 결정에 영향을 주게 된다. 이렇게 각 레이어에서 어떤 feature가 활성화되고, 그것이 어떻게 다음 레이어에 영향을 주는지를 추적하면, 모델이 “지식”, “추론”, “연결” 등을 어떤 방식으로 사용하는지를 circuit-level에서 시각화하고 해석할 수 있게 된다.

- Attribution Graph
하지만 CLT만으로는 feature를 얻는 것까지만 가능하고, 그 feature가 어떻게 출력을 만들어내는지는 파악할 수 없다. 이 문제를 해결하기 위해 사용되는 도구가 Attribution Graph다. 이 그래프는 각 feature가 어떻게 상호작용하며 특정 출력에 영향을 주는지를 시각적으로 보여주는 역할을 한다.

실제로 claude 3.5에서는 "The National Digital Analytics Group (⇪N"이라는 프롬프트에 대해 모델이 "DAG"라는 약어를 자연스럽게 예측할 수 있다. 이때 Attribution Graph를 활용해 Group feature의 활동을 억제하면, 모델은 DAG 대신 ADA, DAM, DAQ와 같은 다른 약어를 생성한다. 이는 해당 그래프가 예측 과정에 실제로 영향을 주고 있음을 보여주는 반례라고 할 수 있다.

- Global Circuit과 Heuristic 추론
이러한 분석은 단일 프롬프트 수준을 넘어서 전체 도메인 단위에서도 활용 가능하다. 도메인 수준 아래서 서로 교차 active 되는 feature들을 취합하면, 도메인 수준의 규칙도 찾아낼 수 있는데, 일반적인 도메인의 경우 관련없는 feature가 너무 많이 활성화되어 해석이 난해하지만, 덧셈 도메인의 경우에는 이례적으로 일관성있는 규칙을 보여주어 모델의 덧셈 특성을 알아볼 수 있었다.
예를 들어 "6 + 9 = 15"라는 연산을 모델이 수행하는 방식은 규칙 기반이라기보다, 수많은 lookup feature들의 조합을 통해 "sum = 15"라는 결과 feature로 이어지는 식이다. computer가 연산하듯 덧셈을 절차적으로 수행하기보다는, 여러가지 heuristic이 composition되어 동작하는 방식이라고 할 수 있디.

- GPT-4.5와 데이터 압축 개념
이런 개념은 OpenAI GPT-4.5이야기를 나눌때 언급하는 압축의 개념과 일맥상통 하는 면이 있다. 해당 영상에서 pretrain이 잘 동작하는 이유를 언급하며, pretrain 과정이 인간이 만든 데이터 압축하는 과정이라고 언급한 바 있다.[4] 좋은 모델은 더 적은 비트로 더 많은 데이터 설명할 수 있는데, 더 효율적인 lookup feature 조합을 만들어 낼 수 있다면 당연히 더 좋은 모델을 만들어낼 수 있을 것이다.
GPT-4.5 케이스를 보면, GPU 수만 늘려서 성능을 높이는 방식은 이미 한계에 가까워졌다는 것을 알수있다. OpenAI도 앞으로는 compute 효율이 아니라 데이터 효율을 높이는 것이 차세대 혁신 방향이라 생각하고 있다.
만약 이런 lookup feature table을 수치화 할 수 있고, 이 table을 효율화하는 방향으로 학습이 가능하다면 해석의 문제를 넘어서서, 더 압축된 형태의 모델을 만들 수 있을 것으로 보인다.

# 레퍼런스
[1] Anthropic - The Urgency of Interpretability https://www.darioamodei.com/post/the-urgency-of-interpretability
[2] https://x.com/slow_developer/status/1918334668669079640?s=12
[3] Circuit Tracing: Revealing Computational Graphs in Language Models https://transformer-circuits.pub/2025/attribution-graphs/methods.html
[4] Pre-Training GPT-4.5 https://www.youtube.com/watch?v=6nJZopACRuQ&ab_channel=OpenAI
[5] Unveiling the Mechanisms of Explicit CoT Training: How Chain-of-Thought Enhances Reasoning Generalization https://arxiv.org/abs/2502.04667
[6] Detecting misbehavior in frontier reasoning models https://openai.com/index/chain-of-thought-monitoring/
[7] Stanford presents global AI trends https://hai.stanford.edu/ai-index/2025-ai-index-report
